samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1430 @@
+}
+
+void SparseImage::free_unused_memory()
+{
+	for (size_t i = 0U; i < virtual_texture.page_table.size(); i++)
+	{
+		virtual_texture.page_table[i].gen_mip_required = false;
+
+		if (virtual_texture.page_table[i].render_required_set.empty() && virtual_texture.page_table[i].valid)
+		{
+			virtual_texture.page_table[i].valid = false;
+			virtual_texture.available_memory_page_indices.push_back(virtual_texture.page_table[i].page_index);
+		}
+	}
+}
+
+void SparseImage::load_least_detailed_level()
+{
+	set_least_detailed_level();
+	compare_mips_table();
+	while (!virtual_texture.texture_block_update_set.empty())
+	{
+		process_texture_blocks();
+		update_and_generate();
+		free_unused_memory();
+	}
+}
+
+void SparseImage::process_stage(enum Stages next_stage)
+{
+	switch (next_stage)
+	{
+		case SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE:
+			calculate_mips_table(current_mvp_transform, on_screen_num_vertical_blocks, on_screen_num_horizontal_blocks, virtual_texture.new_mip_table);
+			this->next_stage = SPARSE_IMAGE_COMPARE_MIPS_TABLE_STAGE;
+			break;
+
+		case SPARSE_IMAGE_COMPARE_MIPS_TABLE_STAGE:
+			compare_mips_table();
+			if (virtual_texture.update_required)
+			{
+				this->next_stage = SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE;
+			}
+			else
+			{
+				this->next_stage = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			frame_counter = 0U;
+			break;
+
+		case SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE:
+			process_texture_blocks();
+			this->next_stage = SPARSE_IMAGE_UPDATE_AND_GENERATE_STAGE;
+			break;
+
+		case SPARSE_IMAGE_UPDATE_AND_GENERATE_STAGE:
+			update_and_generate();
+			free_unused_memory();
+
+			if (frame_counter > 10U)
+			{
+				this->next_stage = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			else if (!transfer_finished)
+			{
+				this->next_stage = SPARSE_IMAGE_PROCESS_TEXTURE_BLOCKS_STAGE;
+			}
+			else
+			{
+				virtual_texture.update_required = false;
+				this->next_stage                = SPARSE_IMAGE_CALCULATE_MIPS_TABLE_STAGE;
+			}
+			break;
+
+		default:
+			break;
+	}
+}
+
+void SparseImage::render(float delta_time)
+{
+	if (!prepared)
+	{
+		return;
+	}
+	if (camera.updated)
+	{
+		update_mvp();
+	}
+	if (color_highlight != color_highlight_mem)
+	{
+		update_frag_settings();
+		color_highlight_mem = color_highlight;
+	}
+
+	process_stage(next_stage);
+
+	ApiVulkanSample::prepare_frame();
+	submit_info.commandBufferCount = 1U;
+	submit_info.pCommandBuffers    = &draw_cmd_buffers[current_buffer];
+	VK_CHECK(vkQueueSubmit(queue, 1U, &submit_info, VK_NULL_HANDLE));
+	ApiVulkanSample::submit_frame();
+}
+
+std::unique_ptr<vkb::VulkanSample> create_sparse_image()
+{
+	return std::make_unique<SparseImage>();
+}
+
+/**
+ * 	@brief Generate the mesh and calculate required mip level for each texture block
+ */
+void SparseImage::calculate_mips_table(glm::mat4 mvp_transform, uint32_t numVerticalBlocks, uint32_t numHorizontalBlocks, std::vector<std::vector<MipBlock>> &mipTable)
```

Comment:
Maybe just return the mip_table, instead of passing in a non-const reference.
And the `mvp_transform` should be passed in by const-reference.

Reply:
Actually, since this function is purely based on `Sparse_Image` meta-data i decided to not use parameters here at all.